File: src/main/java/com/reddit/recommender/ingestion/ScrapeResult.java
package com.reddit.recommender.ingestion;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ScrapeResult {
    private int totalPosts;
    private int totalComments;
    private long startTime;
    private long endTime;
    private List<String> successfulSources;
    private Map<String, String> failedSources; // source -> error message
    private Map<String, Integer> postsBySubreddit;
    private Map<String, Integer> commentsBySubreddit;

    public ScrapeResult() {
        this.successfulSources = new ArrayList<>();
        this.failedSources = new HashMap<>();
        this.postsBySubreddit = new HashMap<>();
        this.commentsBySubreddit = new HashMap<>();
        this.startTime = System.currentTimeMillis();
    }

    // Builder-style methods for fluent usage
    public ScrapeResult addSuccess(String source, int posts, int comments) {
        successfulSources.add(source);
        totalPosts += posts;
        totalComments += comments;

        // Track by subreddit (extract subreddit from source if possible)
        String subreddit = extractSubreddit(source);
        if (subreddit != null) {
            postsBySubreddit.merge(subreddit, posts, Integer::sum);
            if (comments > 0) {
                commentsBySubreddit.merge(subreddit, comments, Integer::sum);
            }
        }

        return this;
    }

    public ScrapeResult addFailure(String source, String error) {
        failedSources.put(source, error);
        return this;
    }

    public ScrapeResult complete() {
        this.endTime = System.currentTimeMillis();
        return this;
    }

    private String extractSubreddit(String source) {
        if (source.startsWith("r/")) {
            return source;
        } else if (source.contains("reddit.com/r/")) {
            // Extract from URL
            String[] parts = source.split("/r/");
            if (parts.length > 1) {
                String sub = parts[1].split("/")[0];
                return "r/" + sub;
            }
        } else if (source.startsWith("keyword:")) {
            return "keyword:" + source.substring(8);
        }
        return source;
    }

    // Getters
    public int getTotalPosts() { return totalPosts; }
    public int getTotalComments() { return totalComments; }
    public long getDurationMs() { return endTime - startTime; }
    public double getPostsPerSecond() {
        long duration = getDurationMs();
        return duration > 0 ? (totalPosts * 1000.0) / duration : 0;
    }
    public List<String> getSuccessfulSources() { return new ArrayList<>(successfulSources); }
    public Map<String, String> getFailedSources() { return new HashMap<>(failedSources); }
    public Map<String, Integer> getPostsBySubreddit() { return new HashMap<>(postsBySubreddit); }
    public Map<String, Integer> getCommentsBySubreddit() { return new HashMap<>(commentsBySubreddit); }

    public String getSummary() {
        return String.format(
                "Scrape completed in %.2f seconds. Posts: %d, Comments: %d, Success: %d, Failed: %d",
                getDurationMs() / 1000.0, totalPosts, totalComments,
                successfulSources.size(), failedSources.size()
        );
    }

    public void printReport() {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("SCRAPE REPORT");
        System.out.println("=".repeat(50));
        System.out.println(getSummary());
        System.out.println("\nPosts per second: " + String.format("%.2f", getPostsPerSecond()));

        if (!postsBySubreddit.isEmpty()) {
            System.out.println("\nPosts by Subreddit:");
            postsBySubreddit.forEach((sub, count) ->
                    System.out.printf("  %-20s: %d posts%n", sub, count));
        }

        if (!failedSources.isEmpty()) {
            System.out.println("\nFailed Sources:");
            failedSources.forEach((source, error) ->
                    System.out.printf("  %-30s: %s%n", source, error));
        }
        System.out.println("=".repeat(50));
    }
}

File: src/main/scala/com/reddit/recommender/ingestion/ScrapeResult.scala
package com.reddit.recommender.ingestion

import scala.collection.mutable

class ScrapeResult {
  private var totalPosts: Int = 0
  private var totalComments: Int = 0
  private var startTime: Long = System.currentTimeMillis()
  private var endTime: Long = 0L

  private val successfulSources = mutable.ListBuffer[String]()
  private val failedSources = mutable.Map[String, String]()
  private val postsBySubreddit = mutable.Map[String, Int]()
  private val commentsBySubreddit = mutable.Map[String, Int]()

  def addSuccess(source: String, posts: Int, comments: Int): ScrapeResult = {
    successfulSources += source
    totalPosts += posts
    totalComments += comments

    val subreddit = extractSubreddit(source)
    postsBySubreddit(subreddit) = postsBySubreddit.getOrElse(subreddit, 0) + posts
    if (comments > 0) {
      commentsBySubreddit(subreddit) = commentsBySubreddit.getOrElse(subreddit, 0) + comments
    }

    this
  }

  def addFailure(source: String, error: String): ScrapeResult = {
    failedSources(source) = error
    this
  }

  def complete(): ScrapeResult = {
    endTime = System.currentTimeMillis()
    this
  }

  private def extractSubreddit(source: String): String = {
    if (source.startsWith("r/")) source
    else if (source.contains("reddit.com/r/")) {
      val parts = source.split("/r/")
      if (parts.length > 1) {
        val sub = parts(1).split("/")(0)
        "r/" + sub
      } else source
    } else if (source.startsWith("keyword:")) source
    else source
  }


  def getTotalPosts: Int = totalPosts
  def getTotalComments: Int = totalComments
  def getDurationMs: Long = endTime - startTime
  def getPostsPerSecond: Double = {
    val duration = getDurationMs
    if (duration > 0) (totalPosts * 1000.0) / duration else 0.0
  }

  def getSuccessfulSources: List[String] = successfulSources.toList
  def getFailedSources: Map[String, String] = failedSources.toMap
  def getPostsBySubreddit: Map[String, Int] = postsBySubreddit.toMap
  def getCommentsBySubreddit: Map[String, Int] = commentsBySubreddit.toMap

  def getSummary: String = {
    f"Scrape completed in ${getDurationMs / 1000.0}%.2f seconds. Posts: $totalPosts%d, Comments: $totalComments%d, Success: ${successfulSources.size}%d, Failed: ${failedSources.size}%d"
  }

  def printReport(): Unit = {
    println("\n" + "=" * 50)
    println("SCRAPE REPORT")
    println("=" * 50)
    println(getSummary)
    printf("\nPosts per second: %.2f\n", getPostsPerSecond)

    if (postsBySubreddit.nonEmpty) {
      println("\nPosts by Subreddit:")
      postsBySubreddit.foreach { case (sub, count) =>
        printf("  %-20s: %d posts%n", sub, count.asInstanceOf[AnyRef])
      }
    }

    if (failedSources.nonEmpty) {
      println("\nFailed Sources:")
      failedSources.foreach { case (source, error) =>
        printf("  %-30s: %s%n", source, error)
      }
    }
    println("=" * 50)
  }
}










